### DISCUSSION TITLE
TP3 - ¿Diferencia entre CAPACIDAD_MAXIMA y capacidad?

### DISCUSSION THREAD

**Discussion by matiaspalleres** [INGCOM-UNRN-P1/cursada-2023#148] [Preguntas y respuestas] [answered]
Eso, estaba repasando el tp3 y me surgio esta duda. Porque en las clases definimos la capacidad maxima como una constante, pero tambien hay un argumento en las funciones que se llama capacidad. No seria mas bien la longitud?

**Comment by vlzJuan** [comment, marked as answer]
Conceptualmente, lo que referenciamos como `CAPACIDAD_MAXIMA` creo que es 'la capacidad en memoria alocada para el array'.

Pensá cada elemento como una cajita. En el TP3 aun no usamos punteros, por lo que solo tenemos "Memoria estática", que solo podemos declarar en nuestro código (Y que no puede cambiar de tamaño durante la ejecución). En este caso, imaginemos que declaro un array de cajitas de enteros:

* `int arreglo[10]={1 , 2, 3 , 4 , 5 , 6 , 7 , 8 , 9 ,10};`

Para este caso, nuestra capacidad máxima seria `10`. Seguramente el `CAPACIDAD_MAXIMA` que tenés está definido como una macro, como por ejemplo:

* `#define CAPACIDAD_MAXIMA 100`

La idea de esa macro es que, al compilar, cada '`CAPACIDAD_MAXIMA`' va a ser reemplazado por un 100. De esa forma, si escribis

* `int arreglo[CAPACIDAD_MAXIMA];`

Es lo mismo que escribir

* `int arreglo[100];`

Es decir: _Le digo al compilador que me guarde `CAPACIDAD_MAXIMA` (100) cajitas que guardan `int`s, una atrás de otra_ Cambiando ese define, cambias de forma rápida y prolija 'la cantidad de cajitas' que puede tener un arreglo.

Por otro lado, hay muchas funciones que hacen referencia a pedir la _capacidad_ como parametro. Esto funciona así:

Si tengo una función 

```C
void mostrar_arreglo(int arreglo[], int capacidad){ //cosas }
```

Le pasas la capacidad como parámetro porque, excepto en `char`s, _no podés saber cuando termina el último elemento de un array_ . Volviendo al ejemplo de cajitas, imaginemos que tenemos `5` cajitas de enteros para un array de enteros `A`, y `5` para un array de enteros `B`, que en memoria están ordenados así:

`|| A[0] || A [1] || A[2] || A[3] || A[4] || B[0] || B[1] || B[2] || B[3] || B[4] ||`

Los `char`s tienen una 'cajita' con el carácter '\0', que se interpreta como el fin del array. En los enteros (`int`), no hay un carácter similar. Entonces, si acá pidieras el dato `A[6]`, estarías accediendo a la 'cajita de enteros' `B[1]`.

La capacidad en la función de mostrar caracteres, por ejemplo, podría usarse de esta forma:

```C
void mostrar_array ( int array[] , int n){
    printf("{");
    for (int i=0; i<n; i++){        // inicializo la variable i, que uso como indice, en 0. Se ejecuta mientras sea menor a n, y aumenta de a 1 por iteracion...
        printf("  %d  ", arreglo[i] );
    }
    printf("}\n");
} 
```

Imprimiría los primeros '`n`' elementos contados desde la cajita '`array[0]`'. Ese '`n`' es tu '`capacidad`' que tenés que poner en las funciones. 

Nota: Incluso si declaras `A` con ‘cantidad de cajitas = `10`’ (es decir, '`int A[10];`', _podes_ si querés dar una 'capacidad' menor, y tu función haría lo que querés hasta esa 'capacidad' que le pusiste. En nuestro ejemplo, si tenés en memoria un A de tamaño 10:

`|| A[0] || A [1] || A[2] || A[3] || A[4] || A[5] || A[6] || A[7] || A[8] || A[9] ||`

Y a la función que te describí antes la llamas como:

```C
mostrar_array( A , 5 );
```

Te mostraría por pantalla hasta el quinto elemento (`A[4]`), en vez de hasta el décimo elemento (`A[9]`). Y es necesario que le digas a la funcion hasta que elemento muestre. 

En el caso de los chars, podés decirle 'Imprimí hasta que detectes el final' (Por ejemplo, cortando si `char[i]=='\0`').  Pero como los enteros no tienen un 'carácter de final', si le decís 'seguí hasta que detectes el final', sin el límite puesto por tu capacidad manual, podrías tranquilamente revisar todas las cajitas del universo y nunca parar (Hasta que el programa se queje y se detenga mostrando un error). Y por eso es necesario que pases tu 'capacidad' como parámetro.

Espero haya servido, saludos

**Reply by matiaspalleres** [reply]
Muchas gracias, tu respuesta me aclaro muchas dudas

**Reply by martinvilu** [reply]
Es lo que mencionábamos como "largo" y "tamaño", pero la explicación de Juan es mas completa.

**Reply by martinvilu** [reply]
No sé que pasó con lo que escribí antes… pero bueno.

> creo que es 'la capacidad en memoria alocada para el array'.

Lo es, es un tamaño para el "bloque" de memoria con el tamaño indicado en el número del `#define`.

Tal vez sea mejor usar la palabra asignada, pero solo para que no se confunda con que la memoria esta _loca_.

Excelente aporte @vlzJuan 


![excelente](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/56625/b2c97647-f020-4c96-9221-08334d5d5908)
